SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularityMesosOfferScheduler.java

```diff
@@ -240,6 +240,71 @@ public SingularityMesosOfferScheduler(MesosConfiguration mesosConfiguration,
     return offerHolders.values();
   }
 
+  private Void calculateScore(
+      Map<String, SingularityOfferHolder> offerHolders,
+      Map<String, RequestUtilization> requestUtilizations,
+      List<SingularityTaskId> activeTaskIds,
+      Map<String, SingularitySlaveUsageWithCalculatedScores> currentSlaveUsagesBySlaveId,
+      Map<String, Integer> tasksPerOfferHost,
+      SingularityTaskRequestHolder taskRequestHolder,
+      Map<String, Double> scorePerOffer,
+      List<SingularityTaskId> activeTaskIdsForRequest,
+      AtomicReference<Throwable> scoringException,
+      SingularityOfferHolder offerHolder) {
+    if (isOfferFull(offerHolder)) {
+      return null;
+    }
+    String slaveId = offerHolder.getSlaveId();
+    Optional<SingularitySlaveUsageWithCalculatedScores> maybeSlaveUsage = Optional.fromNullable(currentSlaveUsagesBySlaveId.get(slaveId));
+
+    if (taskManager.getActiveTasks().stream()
+        .anyMatch(t -> t.getTaskRequest().getDeploy().getTimestamp().or(System.currentTimeMillis()) > maybeSlaveUsage.get().getSlaveUsage().getTimestamp()
+            && t.getMesosTask().getSlaveId().getValue().equals(slaveId))) {
+      Optional<SingularitySlave> maybeSlave = slaveManager.getSlave(slaveId);
+      if (maybeSlave.isPresent()) {
+        CompletableFuture.supplyAsync(() ->
+            usagePoller.collectSlaveUsage(
+              maybeSlave.get(),
+              System.currentTimeMillis(),
+              new ConcurrentHashMap<>(),
+              usageManager.getRequestUtilizations(),
+              new ConcurrentHashMap<>(),
+              new AtomicLong(),
+              new AtomicLong(),
+              new AtomicDouble(),
+              new AtomicDouble(),
+              new AtomicLong(),
+              new AtomicLong()),
+            offerScoringExecutor)
+          .whenComplete((usage, throwable) -> {
+              if (throwable == null && usage.isPresent()) {
+                currentSlaveUsagesBySlaveId.put(slaveId, new SingularitySlaveUsageWithCalculatedScores(
+                    usage.get(),
+                    mesosConfiguration.getScoreUsingSystemLoad(),
+                    getMaxProbableUsageForSlave(activeTaskIds, requestUtilizations, offerHolders.get(slaveId).getSanitizedHost()),
+                    mesosConfiguration.getLoad5OverloadedThreshold(),
+                    mesosConfiguration.getLoad1OverloadedThreshold(),
+                    usage.get().getTimestamp()
+                ));
+              } else {
+                throw new RuntimeException(throwable);
```

Comment:
Where is the handling for this runtime exception? We currently aren't calling a get or join for the future created here which causes two issues for us:
- calculateScore below may end up being called before the metrics recollection has a chance to run
- If a RuntimeException is thrown here it is lost to us since there it will not propagate out of the future and no catch block currently logs it